// @ionic-enterprise/auth/dist/esm/SecureWebView.js
export class SecureWebView {
    isAvailable(callback) {
        var errorHandler = function errorHandler(error) {
            // An error has occurred while trying to access the
            // SecureWebView native implementation, most likely because
            // we are on an unsupported platform.
            callback(false);
        };
        cordova.exec(callback, errorHandler, 'SecureWebView', 'isAvailable', []);
    }
    // options:
    //  url - url to display
    //  webView - for iOS which webview to display, if possible. By default we use the newest one available for OS version
    //      - ASWebAuth - ASWebAuthenticationSession (avaialble starting in iOS 12)
    //      - SFAuth - SFAuthenticationSession (available starting in iOS 11)
    //      - SFSafari - SFSafariViewController (available starting in iOS 9)
    //      - MobileSafari - Mobile Safari (pre-iOS 8)
    show(options, onSuccess, onError) {
        options = options || {};
        if (!options.hasOwnProperty('animated')) {
            options.animated = true;
        }
        cordova.exec(onSuccess, onError, 'SecureWebView', 'show', [options]);
    }
    hide(onSuccess, onError) {
        cordova.exec(onSuccess, onError, 'SecureWebView', 'hide', []);
    }
    getViewHandlerPackages(onSuccess, onError) {
        cordova.exec(onSuccess, onError, 'SecureWebView', 'getViewHandlerPackages', []);
    }
    useCustomTabsImplementation(packageName, onSuccess, onError) {
        cordova.exec(onSuccess, onError, 'SecureWebView', 'useCustomTabsImplementation', [packageName]);
    }
    connectToService(onSuccess, onError) {
        cordova.exec(onSuccess, onError, 'SecureWebView', 'connectToService', []);
    }
    warmUp(onSuccess, onError) {
        cordova.exec(onSuccess, onError, 'SecureWebView', 'warmUp', []);
    }
    mayLaunchUrl(url, onSuccess, onError) {
        cordova.exec(onSuccess, onError, 'SecureWebView', 'mayLaunchUrl', [url]);
    }
}
export const IonicSecureWebView = new SecureWebView();
